### PR TITLE
[cxx] Rework C++ linkage to C99 trunc, isnan, isinf, etc. and some cleanup.

### PR DESCRIPTION
**Pull request by jaykrell**
Android was giving quite some problems otherwise, and WebAssembly less so.

### CODE REVIEW

**Comment by jaykrell** [comment]
@monojenkins build OS X x64 Android SDK C++
@monojenkins build OS X x64 iOS SDK C++
@monojenkins build Linux WebAssembly C++

**Comment by jaykrell** [comment]
This is meant to fix e.g.

https://jenkins.mono-project.com/job/test-mono-pull-request-amd64-osx-products-sdks-android-cxx/12/parsed_console/log.html

unction 'void interp_exec_method_full(InterpFrame*, ThreadContext*, guint16*, MonoException*, int, InterpFrame*)':
/Users/builder/jenkins/workspace/test-mono-pull-request-amd64-osx-products-sdks-android-cxx/mono/mini/interp/interp.c:4885:32: error: 'isfinite' was not declared in this scope
    if (!isfinite(sp [-1].data.f))
                                ^

and just be a more complete reliable maintainable technique.
Albeit a little larger and slower perhaps.

Alternatively, adding using std::isinfinite or a mono_isinfinite to interp.c should progress.

**Comment by jaykrell** [comment]
The hacks.h file makes this all kinda poor defined.

**Comment by jaykrell** [comment]
@monojenkins build OS X x64 Android SDK C++
@monojenkins build OS X x64 iOS SDK C++
@monojenkins build Linux WebAssembly C++

**Comment by jaykrell** [comment]
@monojenkins build OS X x64 Android SDK C++
@monojenkins build OS X x64 iOS SDK C++
@monojenkins build Linux WebAssembly C++

**Comment by jaykrell** [comment]
@monojenkins build

**Comment by jaykrell** [comment]
@monojenkins build OS X x64 Android SDK C++
@monojenkins build OS X x64 iOS SDK C++
@monojenkins build Linux WebAssembly C++

**Comment by jaykrell** [comment]
@monojenkins build OS X x64 Android SDK C++
@monojenkins build OS X x64 iOS SDK C++
@monojenkins build Linux WebAssembly C++

**Comment by jaykrell** [comment]
@monojenkins build OS X x64 C++

**Comment by jaykrell** [comment]
Removal of hacks.h makes this much clearer.
The isunordered hack could be preserved if really needed.
The define of iso99 seems unnecessary but could be preserved as well.

**Comment by jaykrell** [comment]
@monojenkins build OS X x64 Android SDK C++
@monojenkins build OS X x64 iOS SDK C++
@monojenkins build Linux WebAssembly C++

**Comment by jaykrell** [comment]
@monojenkins build OS X x64 Android SDK C++
@monojenkins build OS X x64 iOS SDK C++
@monojenkins build Linux WebAssembly C++

File: mono/mini/mini-runtime.c [hunk 1]
```diff
@@ -4657,7 +4657,7 @@ register_icalls (void)
 		register_icall (mono_fload_r4, "mono_fload_r4", "double ptr", FALSE);
 		register_icall (mono_fstore_r4, "mono_fstore_r4", "void double ptr", FALSE);
 		register_icall (mono_fload_r4_arg, "mono_fload_r4_arg", "uint32 double", FALSE);
-		register_icall (mono_isfinite, "mono_isfinite", "uint32 double", FALSE);
+		register_icall (mono_isfinite, "mono_jit_icall_isfinite", "uint32 double", FALSE);
```
Comment (lambdageek) [inline mono/mini/mini-runtime.c]: Should be `register_icall (mono_jit_icall_isfinite, "mono_jit_icall_isfinite", ...)`

Comment (jaykrell) [reply mono/mini/mini-runtime.c]: Good catch.

Comment (jaykrell) [reply mono/mini/mini-runtime.c]: I'd also be open to a new naming scheme. mono_math_isinf(), mon_c99_isinf, or something. I was reluctant to rename this function, but discovered it late, and think I picked a pretty good name for the others.

We could also expose mono_isinfinite_double here, the new function to replace the old, but they seem just ever so slightly different in intent -- icall vs. c99 wrappers.

**Comment by jaykrell** [comment]
@monojenkins build OS X x64 Android SDK C++
@monojenkins build OS X x64 iOS SDK C++
@monojenkins build Linux WebAssembly C++

**Comment by jaykrell** [comment]
@monojenkins build OS X x64 Android SDK C++
@monojenkins build OS X x64 iOS SDK C++
@monojenkins build Linux WebAssembly C++

**Comment by jaykrell** [comment]
@monojenkins build Linux x64

**Comment by jaykrell** [comment]
@monojenkins build OS X x64

**Comment by jaykrell** [comment]
https://github.com/mono/mono/issues/10769

File: mono/utils/mono-math-c.c [hunk 2]
```diff
@@ -0,0 +1,22 @@
+// This file bridges C++ to C runtime math functions
+// without any chance of a C++ library dependency.
+// In time it can probably be removed.
+#include "config.h"
+#include "glib.h"
+#define MONO_MATH_DECLARE_ALL 1
+#include "mono-math.h"
+
+#if defined (__cplusplus) && !defined (_MSC_VER)
```
Comment (jaykrell) [inline mono/utils/mono-math-c.c]: fixme: Project file work to compile as C also for Visual C++? But it doesn't really matter. This work is mostly for Android and WebAssembly. Mac/Windows/Linux were all building ok previously. iOS uncertain.
What put me over the edge to not just keep using "using" is the symbols imported on Linux in a toy example -- mangled isinf or such did appear. This should avoid that.
Eventually maybe the inline functions in mono-math.h can just call std::foo but this approach is definitely a little safer, for the cost of roughly an instruction per call to these functions.

**Comment by jaykrell** [comment]
https://github.com/mono/mono/issues/10797

**Comment by jaykrell** [comment]
@monojenkins build OS X x64

**Comment by jaykrell** [comment]
@monojenkins build Linux ARMv5

**Comment by jaykrell** [comment]
@monojenkins build Linux ARMv5

**Comment by jaykrell** [comment]
@monojenkins build Linux ARMv5

**Comment by jaykrell** [comment]
@monojenkins build Linux ARMv5

**Pull request merged**